### DISCUSSION TITLE
Ideas for more meaningful 'Jitter LED' in Client Dialog

### DISCUSSION THREAD

**Discussion by passing** [jamulussoftware/jamulus#1815] [UI Related]
Right now the 'Jitter' LED has just 2 states:
- Red for any number of errors within a 300ms interval:
  - Client Jitter Buffer failure https://github.com/jamulussoftware/jamulus/blob/master/src/client.cpp#L1145
  - "audio" / "generic" errors reading audio packets https://github.com/jamulussoftware/jamulus/blob/master/src/socket.cpp#L281-L282
- Green for none of these

I tried to make this more meaningful and modified the code to actually count the different errors (counters reset in a defined interval just like the status) and show the counted values next to the LED (which then can also go yellow for a low number of errors):

![image](https://user-images.githubusercontent.com/8142186/120560544-29ca4580-c403-11eb-8dc0-abbf60eb4a65.png) ![image](https://user-images.githubusercontent.com/8142186/120560510-1dde8380-c403-11eb-88bd-e9e924f6b18d.png) ![image](https://user-images.githubusercontent.com/8142186/120560558-3484da80-c403-11eb-9f06-37d15c4b6a5f.png) ![image](https://user-images.githubusercontent.com/8142186/120560490-17e8a280-c403-11eb-8f10-9a868c1377d5.png) 

I think it at least looks nice and may be helpful in some cases, still for many users this may not give any benefit.
So I would like to get some feedback if that (or part of that, or something similar) is desirable, before I put more effort into nice and shiny code ;)

**Comment by gene96817** [comment]
Sorry, I am a dunce reading the code from the link you provided.
What are the conditions for buffer failure? It seems that buffer underrun or overrun counts as failures.
What is the meaning for Jitter =x/y?
Thanks....

**Reply by ghost** [reply]
I think the LEDs are good enough. Less is better, more is burdensome.

**Reply by passing** [reply]
@gene96817 
> What is the meaning for Jitter =x/y?

I assume that
- x is the number of buffer underruns in the Client Jitter Buffer
- y is the number of buffer overruns in the Client Jitter Buffer

but I am also not too experienced in the Jamulus codebase, so this may not be correct

**Comment by gilgongo** [comment]
When you say "may be helpful in some cases" what would somebody do if (for example) the errors were 5/0 as opposed to 0/361? If counting the errors get you closer to a reliable action to take in order to fix the errors, then I'd be more interested in showing a call to action about that ("Make your buffers bigger"?)

**Reply by passing** [reply]
That's exactly what I hope to clarify here as I just don't know.

Having looked a bit more at the code now, I think there are actually just 2 relevant cases that cause a red LED:
1. CChannel::GetData returns GS_BUFFER_UNDERRUN
which obviously seems to be a buffer underrun
1. CChannel::PutAudioData returns PS_AUDIO_ERR
which I assume is a buffer overrun (?)

And I now just counted and displayed these separately.
I'm especially unsure about the 2nd case: That counter goes quite high on some servers, independent of the buffer settings

**Reply by gene96817** [reply]
With the explanations, it seems that x/y should be errors of any kind, with x = inbound errors and y = outbound errors.
However, outbound errors (specifically, buffer overrun) are of limited utility. This would be a delay in the Jamulus internal data flow. If the path has multiple hops, the x/y is not actionable since we still don't know what hop contributes to the error.

**Reply by passing** [reply]
the display could just be reduced to a single number which would show the buffer underrun count as a positive number and the buffer overrun as a negative number. then the tooltip could explain something like:
- small positive number are usually acceptable, for a bigger number the buffer should be increased when the audio quality is getting bad
- negative numbers for a longer time indicate an issue in the client or server

**Comment by ghost** [comment]
Jamulus is supposed to adhere to the KISS principle. Numbers beside LEDs add to confusion and code bloat and cpu load and battery draining and global warming. The LEDs are sufficiently sufficient.

**Reply by passing** [reply]
got your opinion already after the first comment! I still want to get some more.
If this turns out to really not provide any value, I'm totally fine with discarding it.

**Comment by passing** [comment]
I agree to the feedback that - if at all - the additional information should be kept at a minimum. So my proposal would be just a single number counting the underruns and showing overruns as a negative number when there are no underruns (usually doesn't happen):

![image](https://user-images.githubusercontent.com/8142186/120890688-b7c14e80-c604-11eb-9f79-b907a6a2c7ce.png)![image](https://user-images.githubusercontent.com/8142186/120890692-babc3f00-c604-11eb-8cae-112920ef7d51.png)

Also noticed that a common problem for buffer underruns is using wifi (which of course you shoudn't), so I'd include that in the tooltip

draft for information to be added to the tooltip:
- small numbers of jitter buffer errors are usually acceptable
- when audio quality is bad while the counter shows a bigger value (to be defined)
  - ensure a stable network connection: use a network cable instead of wifi
  - otherwise try increasing the local jitter buffer
- negative numbers for a longer time indicate an issue in the client or server

I have committed my modifications on a branch here: https://github.com/passing/jamulus/tree/jitter_failure_counter
It would be nice if you could try it out yourselves before judging on it. It makes much more sense to see this in action instead of on a screenshot. ( @jamulussoftware/maindevelopers , it would be nice if one of you could try it as well )

**Reply by ann0see** [reply]
Hmm. I‘m still not sure if this adds more value. What about extending the analyser console with more useful information as it shows now and also add some code from nefarius repo?

**Reply by gene96817** [reply]
What causes the buffer underrun count to increment? Does it increment during longer periods of undnerrun?
A buffer underrun effectively inserts a time offset (delay) to the audio stream. This offset is cumulative. The only way to eliminate this offset is to flush the late packets. Does the code have the ability to flush late (or very late) packets?
Since the code can detect out of order packets, the packets are numbered. This makes it possible to know how many packets need to be flushed.

**Reply by gilgongo** [reply]
Jamulus is I hope mainly about playing music, not signal analysis. So for something to be on the main screen as opposed to another screen, there needs to be some clear rationale for it in terms of playing music. For example, the sliders are there for when you can't hear someone well enough, or they are too loud. So you adjust them. The level meters are there so that you can tell which channels are active and by how much so that you can decide to mute, pan, turn them down, etc.

Is there any "musical action" you can take on the info displayed by the jitter/ping status? Indeed, removing it all has been discussed in the past by Volker and others on just these grounds.  So as @ann0see says - is about the analyser console because playing and analysing your sound are separate things. 

In fact, @passing your answer to my question above seems to confirm this: https://github.com/jamulussoftware/jamulus/discussions/1815#discussioncomment-823250 In general, information presented by UIs need to have clearly actionable reasons to be there.

**Reply by passing** [reply]
> What causes the buffer underrun count to increment?

@gene96817 the only way I found by accident to get the buffer overrun up (i guess you wanted to know about that?!) was a lower frames/second setting in jackctl than the hardware could handle (resulting in no sound being output).

**Reply by gene96817** [reply]
@gilgongo    I am interested in your opinion of my observation. Think of it as an extension of correcting out-of-order packets. Nothing needs to be seen at the UI. This affects sound quality under jitter conditions.

**Reply by gene96817** [reply]
>  I found by accident to get the buffer overrun up (i guess you wanted to know about that?!) was a lower frames/second setting in jackctl than the hardware could handle (resulting in no sound being output).

I was asking about underrun. As to overrun, yes if you have something in the data path that has a low frames/second, you will have an overrun condition at the bottleneck that causes packets to be dropped. With enough dropped packets, there will be no sound output.

**Comment by passing** [comment]
> for something to be on the main screen as opposed to another screen, there needs to be some clear rationale for it in terms of playing music

@gilgongo well my intention is not to have this on the main screen. Just the small section dedicated to showing technical metrics is already there:
![image](https://user-images.githubusercontent.com/8142186/120900083-48168800-c633-11eb-945a-e1989bd0c990.png)
In that section, the LEDs alone just tell: green=good, yellow=not so good, red=bad. And there's also numeric values for Delay and Ping. So I just thought: there's a huge gap next to "Jitter", why not also put the number there that quantifies that Jitter LED. :smile: 

Anyway, part of my change is to have yellow color on the Jitter LED as well, as right now a single underrun makes it red already. And I think red should only be shown if something is really bad.

> In general, information presented by UIs need to have clearly actionable reasons to be there.

I think that still makes sense:
- For a small number of buffer failures (yellow): no action is required
- For a larger number of buffer failures (red):
   - same hints as now (jitter buffer, sound card buffer, internet rate, cpu) and one more: replace wifi with wired connection

**Reply by gilgongo** [reply]
> Just the small section dedicated to showing technical metrics is already there:

Sorry, I meant that **it is** on the main screen. 

The main screen is about playing, not analysing. You wouldn't be able to do things like replace wifi with wired while you're playing, and ultimately its your own ears that are the judge of whether you need to start investigating anything. I'd say keep telemetry off that screen unless we can isolate a specific state in Jamulus to a specific call to action in the same way as "too loud = turn it down". But we can't.

BTW there's a related discussion on the issue of showing metrics generally here: https://github.com/jamulussoftware/jamulus/issues/762

**Reply by passing** [reply]
> The main screen is about playing, not analysing

@gilgongo it is just confusing to me that **all** the analytics features have just been moved from the settings dialog to the main screen with 3.8.0 (https://github.com/jamulussoftware/jamulus/pull/1762). But it seems it is actually not wanted to have them there. And if that's the case, it would be good to know what the desired state would actually be, to be able to work towards that.

> and ultimately its your own ears that are the judge of whether you need to start investigating anything

You can hear certain issues, but I don't think you can usually hear, which user causes them. And then you usually need the numbers. In the sessions I have, there's usually a few users that are barely able to setup their system. When there are issues it is really helpful to have the important metrics at hand easily. Being able to ask "hey everyone, look at the number at the left of your Jamulus Window next to 'Delay', does anybody have something higher than 50 there?" is just a bless with 3.8.0! Having to explain multiple times how to open the settings dialog to find that number was just a pain before.

**Reply by dcorson-ticino-com** [reply]
>it is just confusing to me that all the analytics features have just been moved from the settings dialog to the main screen

@passing Hi Martin, 
Moving the ping and delay to the main window was not done because we thought it is better after a long decision porocess.  It was done to reduce the frequency of a problem on relatively new Macs that comes with updating values on windows open in the background, such as happens when people have the Settings window open to see their ping time.  Apparently there are many people who do this.  
The problem appears to be caused by a problem in the Mac Qt library or Mac library, it does not seem to happen on Linux and Win10.  And the same program on older Mac versions runs without a problem.  We still don't know the root cause, even after some intensive debugging. This was the only action that we found that improved things.
We continue to stalk this bug, but at this moment do not have a hot clue.
Once we have solved this bug we may reconsider the placement of the ping and delay, but as you say, there is quite a faction that prefer it as it now is.

**Reply by gilgongo** [reply]
@passing 

> But it seems it is actually not wanted to have them there. 

It's the consensus to have them where they are, but my point is that until these metrics are much more fine-grained, there is no clearly actionable effect of having them there. They offer an illusion of being in control. In reality, if you have a high ping, or jitter or overall delay, you can do nothing beyond the implementation of some general best practice (buying better equipment, not using wi-fi, increasing your buffers, maybe checking for a CPU hog).

@dcorson-ticino-com point notwithstanding, rather than showing flashing lights and numbers, it would be far better to work on some kind of call to action, or even automatic response, similar to that of the recently introduced feedback detection, for example.

> "hey everyone, look at the number at the left of your Jamulus Window next to 'Delay', does anybody have something higher than 50 there?"

... and so... what?

**Reply by passing** [reply]
> > "hey everyone, look at the number at the left of your Jamulus Window next to 'Delay', does anybody have something higher than 50 there?"
> 
> ... and so... what?

then we know who is the problem, and suggest that everybody else turns the volume of his channel down. So everybody else can continue with a better experience in the current session. The one with the bad ping can then still participate 'passively' and we offer to help analyzing his/her setup at a later date so it can be improved for the next time.

**Reply by gilgongo** [reply]
OK, so that's the sort of "indication to action" thing I'm talking about: if X then do Y. But you don't need Ping **and** Delay **and** Jitter LED, **and** have numbers next to them to achieve that - indeed you could just automate it. We show a "muted" icon today if a channel cannot hear you. Show a "disrupted" icon next to the channel that's disrupting if that's all that matters at the end of the day. Why all the UI invasion?

**Reply by passing** [reply]
> indeed you could just automate it.

I agree that would be the ultimate goal, but hard to achieve

I don't agree that providing more information on the main UI is a problem in general. While using Jamulus, I just play music and don't look at the UI almost all the time. But when I want to adjust something or see if there are issues, I prefer having everything at hand quickly (and don't mind seeing more information than I need)
But that's just my personal opinion and I'm fine with the different feedback I got - so will stop this initiative here.

**Reply by gilgongo** [reply]
In so far as channel mute status is being sent to the server today, I assume a "disrupted" indicator would not be too hard (although I'd guess it would need a protocol change). Given that it seems to solve the problem behind the problem, as it were, could you not shift the initiative in that direction? You have articulated a problem after all.  I'm simply objecting to the execution of proposed solution.

**Reply by passing** [reply]
if I understand you right, you are talking about distributing connection-quality related information via the server to the clients? Then I guess that specific requirement should be part of https://github.com/jamulussoftware/jamulus/issues/961

But that's probably way outside of my still quite small comfort zone in regards to C++

**Reply by gilgongo** [reply]
It depends what you mean by "connection quality". If it's stuff like jitter buffer numbers and ping stats, then forget that for the reasons I've stated (they're basically useless). But if it's a simple "something is probably perturbing the signal" flag the yes, because that's actionable and seems useful in the way you have described.
